# **ĐỒ ÁN MÔN HỌC: ỨNG DỤNG PHÂN TÍCH DỮ LIỆU THÔNG MINH**

## **Thực hiện:**
|<font size=4> Tên </font>|<font size=4> MSSV </font>|
|-------------------------|----------------------------------|
|<font size=4> Nguyễn Đình Hoàng Nguyên </font>|<font size=4> 19120606 </font>|

# **Giai đoạn 2: Tiền xử lý dữ liệu**

## **2.0. Tải dữ liệu từ github**

In [1]:
!wget https://raw.githubusercontent.com/HoangNguyen31/Project_Applications-of-Intelligent-Data-Analysis/main/data/apartments.json

--2024-03-04 10:00:44--  https://raw.githubusercontent.com/HoangNguyen31/Project_Applications-of-Intelligent-Data-Analysis/main/data/apartments.json
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 15028440 (14M) [text/plain]
Saving to: ‘apartments.json’

apartments.json     100%[===================>]  14.33M  --.-KB/s    in 0.07s   

2024-03-04 10:00:45 (195 MB/s) - ‘apartments.json’ saved [15028440/15028440]



## **2.1. Import thư viện**

In [2]:
import pandas as pd
import numpy as np

import warnings
warnings.filterwarnings("ignore")

## **2.2. Đọc dữ liệu**

In [3]:
df = pd.read_json("apartments.json")
df.head()

ad_id    list_id  category    category_name  region     region_name  \
0  155094011  114294787      1010  Căn hộ/Chung cư      13  Tp Hồ Chí Minh   
1  154943228  114163716      1010  Căn hộ/Chung cư      13  Tp Hồ Chí Minh   
2  154792197  114032647      1010  Căn hộ/Chung cư      13  Tp Hồ Chí Minh   
3  152808559  112328722      1010  Căn hộ/Chung cư      13  Tp Hồ Chí Minh   
4  154792211  114032660      1010  Căn hộ/Chung cư      13  Tp Hồ Chí Minh   

   area          area_name     ward                    ward_name  ... rooms  \
0   119  Thành phố Thủ Đức  11221.0  Phường An Khánh (Quận 2 cũ)  ...     1   
1   119  Thành phố Thủ Đức   9341.0    Phường An Phú (Quận 2 cũ)  ...     1   
2   119  Thành phố Thủ Đức  11221.0  Phường An Khánh (Quận 2 cũ)  ...     1   
3   109    Quận Bình Thạnh   9284.0                    Phường 25  ...     3   
4   112      Quận Tân Bình   9301.0                    Phường 12  ...     2   

   toilets longitude  latitude  type  type_name  owner  deposit  \
0      1.0  106.7304   10.7929     u   Cho thuê  False      NaN   
1      1.0  106.7512   10.8073     u   Cho thuê  False      NaN   
2      1.0  106.7304   10.7929     u   Cho thuê  False      NaN   
3      2.0  106.7160   10.8059     u   Cho thuê  False      NaN   
4      2.0  106.6523   10.8030     u   Cho thuê  False      NaN   

  price_million_per_m2       price  
0             0.216667   6500000.0  
1             0.250000  10000000.0  
2             0.240000   6000000.0  
3             0.136364  15000000.0  
4             0.162162  12000000.0  

[5 rows x 24 columns]

## **2.3. Kích thước dữ liệu, ý nghĩa các dòng và cột thuộc tính**

- Kích thước tập dữ liệu

In [4]:
rows, columns = df.shape
print(f"+ Tập dữ liệu có {rows} dòng dữ liệu.")
print(f"+ Tập dữ liệu có {columns} cột thuộc tính.")

+ Tập dữ liệu có 19829 dòng dữ liệu.
+ Tập dữ liệu có 24 cột thuộc tính.


- Ý nghĩa các dòng dữ liệu: Mỗi dòng dữ liệu chứa thông tin của căn hộ, chung cư cho thuê.

- Ý nghĩa các cột thuộc tính:

In [5]:
df.columns

Index(['ad_id', 'list_id', 'category', 'category_name', 'region',
       'region_name', 'area', 'area_name', 'ward', 'ward_name', 'street_name',
       'company_ad', 'apartment_type', 'size', 'rooms', 'toilets', 'longitude',
       'latitude', 'type', 'type_name', 'owner', 'deposit',
       'price_million_per_m2', 'price'],
      dtype='object')

| Thuộc tính           | Mô tả                               |
|----------------------|-------------------------------------|
| ad_id                | ID của quảng cáo                    |
| list_id              | ID trong danh sách quảng cáo        |
| category             | Mã bất động sản                     |
| category_name        | Tên bất động sản                    |
| region               | Mã tỉnh                             |
| region_name          | Tên tỉnh                            |
| area                 | Mã quận                             |
| area_name            | Tên quận/huyện                      |
| ward                 | Mã phường                           |
| ward_name            | Tên phường                          |
| street_name          | Tên đường                           |
| company_ad           | Là quảng cáo của công ty hay không? |
| apartment_type       | Loại căn hộ, chung cư               |
| size                 | Kích thước phòng                    |
| rooms                | Số phòng                            |
| toilets              | Số toilet                           |
| longitude            | Kinh độ                             |
| latitude             | Vĩ độ                               |
| type                 | Mã loại thuê                        |
| type_name            | Tên loại thuê                       |
| owner                | Có phải chủ nhà không?              |
| deposit              | Tiền đặt cọc                        |
| price_million_per_m2 | Giá trên mỗi mét vuông (VNĐ)        |
| price                | Giá phòng(VNĐ)                      |

## **2.4. Kiểu dữ liệu hiện tại của mỗi cột là gì? Có cột thuộc tính nào có kiểu dữ liệu không phù hợp không?**

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19829 entries, 0 to 19828
Data columns (total 24 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   ad_id                 19829 non-null  int64  
 1   list_id               19829 non-null  int64  
 2   category              19829 non-null  int64  
 3   category_name         19829 non-null  object 
 4   region                19829 non-null  int64  
 5   region_name           19829 non-null  object 
 6   area                  19829 non-null  int64  
 7   area_name             19829 non-null  object 
 8   ward                  19828 non-null  float64
 9   ward_name             19821 non-null  object 
 10  street_name           19818 non-null  object 
 11  company_ad            19829 non-null  bool   
 12  apartment_type        19829 non-null  object 
 13  size                  19829 non-null  float64
 14  rooms                 19829 non-null  int64  
 15  toilets            

Nhận xét:
- Tập dữ liệu có:
  - 2 thuộc tính dạng `bool`.
  - 8 thuộc tính dạng `float`.
  - 6 thuộc tính dạng `int`.
  - 8 thuộc tính dạng `object`.
- Tất cả các kiểu dữ liệu đều phù hợp.

## **2.5. Tình trạng lặp dữ liệu**

In [7]:
duplicated_rows = df.duplicated().sum()
print(f"Dữ liệu trên có {duplicated_rows} dòng bị lặp.")

Dữ liệu trên có 0 dòng bị lặp.


## **2.6. Với mỗi cột số, các giá trị được phân bố như thế nào?**

In [8]:
numeric_df = df.select_dtypes(include=["number"])
nume_col_df = numeric_df.describe()
nume_col_df.loc["missing_count"] = numeric_df.isna().sum()
nume_col_df.loc["missing_ratio"] = (numeric_df.isna().sum() / numeric_df.shape[0]) * 100
nume_col_df

ad_id       list_id  category        region  \
count          1.982900e+04  1.982900e+04   19829.0  19829.000000   
mean           1.535889e+08  1.130096e+08    1010.0     12.998891   
std            3.340528e+06  2.797164e+06       0.0      0.033291   
min            5.430738e+07  3.571098e+07    1010.0     12.000000   
25%            1.536657e+08  1.130652e+08    1010.0     13.000000   
50%            1.548068e+08  1.140451e+08    1010.0     13.000000   
75%            1.549779e+08  1.141940e+08    1010.0     13.000000   
max            1.551357e+08  1.143311e+08    1010.0     13.000000   
missing_count  0.000000e+00  0.000000e+00       0.0      0.000000   
missing_ratio  0.000000e+00  0.000000e+00       0.0      0.000000   

                       area          ward          size         rooms  \
count          19829.000000  19828.000000  19829.000000  19829.000000   
mean             108.802007   9397.734113     55.307949      1.525443   
std                6.925971    309.641363    161.106571      0.728481   
min               96.000000   9217.000000      6.000000      1.000000   
25%              102.000000   9285.000000     35.000000      1.000000   
50%              109.000000   9332.000000     50.000000      1.000000   
75%              113.000000   9447.000000     70.000000      2.000000   
max              119.000000  11222.000000  20000.000000     11.000000   
missing_count      0.000000      1.000000      0.000000      0.000000   
missing_ratio      0.000000      0.005043      0.000000      0.000000   

                    toilets     longitude      latitude       deposit  \
count          19828.000000  19826.000000  19826.000000  8.288000e+03   
mean               1.370890    106.691338     10.782634  1.351341e+07   
std                0.552051      0.048022      0.039000  6.402050e+07   
min                1.000000    106.509620     10.663900  1.000000e+00   
25%                1.000000    106.661300     10.755900  5.000000e+06   
50%                1.000000    106.688000     10.789600  7.500000e+06   
75%                2.000000    106.720200     10.805800  1.800000e+07   
max                7.000000    106.856000     10.891700  5.500000e+09   
missing_count      1.000000      3.000000      3.000000  1.154100e+04   
missing_ratio      0.005043      0.015129      0.015129  5.820263e+01   

               price_million_per_m2         price  
count                  19828.000000  1.982800e+04  
mean                       0.270693  1.697045e+07  
std                        4.363408  3.585843e+08  
min                        0.000400  5.000000e+05  
25%                        0.135714  5.800000e+06  
50%                        0.171429  7.500000e+06  
75%                        0.211538  1.100000e+07  
max                      338.709677  3.000000e+10  
missing_count              1.000000  1.000000e+00  
missing_ratio              0.005043  5.043119e-03

Nhận xét:
- Các cột `category (Mã bất động sản)` và `region (Mã tỉnh)` đều chỉ có 1 giá trị duy nhất lần lượt là `1010` và `13`. Ta cần cân nhắc loại bỏ 2 cột này.
- Các cột thuộc tính còn lại có một số giá trị bị thiếu, đặc biệt là ở các cột như `ward`, `toilets`, `longitude`, `latitude`, và `deposit`. Cần xem xét các biện pháp để xử lý dữ liệu bị thiếu này.
- Các cột như `deposit`, `price_million_per_m2`, `price` đòi hỏi kiểm tra kỹ lưỡng để loại bỏ các giá trị ngoại lai.

## **2.7. Với mỗi cột phân loại, các giá trị được phân bổ như thế nào?**

In [9]:
object_df = df.select_dtypes(include="object")
object_col_df = object_df.describe()
object_col_df.loc["missing_count"] = object_df.isna().sum()
object_col_df.loc["missing_ratio"] = (object_df.isna().sum() / object_df.shape[0]) * 100
object_col_df

category_name     region_name          area_name ward_name  \
count                    19829           19829              19829     19821   
unique                       1               1                 20       129   
top            Căn hộ/Chung cư  Tp Hồ Chí Minh  Thành phố Thủ Đức  Phường 2   
freq                     19829           19829               3480       865   
missing_count                0               0                  0         8   
missing_ratio              0.0             0.0                0.0  0.040345   

                   street_name apartment_type   type type_name  
count                    19818          19829  19829     19829  
unique                    2384              6      1         1  
top            Nguyễn Văn Linh       Chung cư      u  Cho thuê  
freq                       510          11531  19829     19829  
missing_count               11              0      0         0  
missing_ratio         0.055474            0.0    0.0       0.0

Nhận xét:
- Các cột thuộc tính `category_name`,	`region_name`, `type`, `type_name` đều chỉ có một giá trị duy nhất. Ta cũng cần cân nhắc loại bỏ các cột này.
- Các cột thuộc tính còn lại có một số giá trị bị thiếu. Cần xem xét các biện pháp để xử lý dữ liệu bị thiếu này

## **2.8. Loại bỏ các cột thuộc tính không có ý nghĩa**

In [10]:
column_to_drop = ["category", "region", "category_name", "region_name", "type", "type_name"]
df.drop(column_to_drop, axis=1, inplace=True)

## **2.8. Tình trạng khuyết dữ liệu**

In [11]:
null_df = pd.DataFrame(df.isnull().sum(), columns=["Số mẫu bị thiếu"])
null_df["Tỉ lệ bị thiếu"] = np.round(df.isnull().sum() / df.shape[0] * 100, 2)
null_df.index.name = "Thuộc tính"
null_df = null_df.sort_values(by="Tỉ lệ bị thiếu", ascending=False)
null_df.loc[null_df["Tỉ lệ bị thiếu"] > 0]

Số mẫu bị thiếu  Tỉ lệ bị thiếu
Thuộc tính                                           
deposit                         11541           58.20
street_name                        11            0.06
ward_name                           8            0.04
latitude                            3            0.02
longitude                           3            0.02
price                               1            0.01
toilets                             1            0.01
price_million_per_m2                1            0.01
ward                                1            0.01

Nhận xét:
- Thuộc tính `deposit` là thuộc tính thiếu nhiều nhất, khoảng `58.20%` dữ liệu bị thiếu.
- Các thuộc tính thiếu còn lại có số lượng dữ liệu bị thiếu không đáng kể.

## **2.9. Xử lý tình trạng khuyết dữ liệu**

- Thuộc tính `deposit` ta sẽ điền giá trị `0` tượng trưng cho không cần đóng tiền cọc trước.

In [12]:
df["deposit"].fillna(0, inplace=True)

- Thuộc tính `ward_name` sẽ được điền dựa trên theo mã của `ward`.

In [13]:
df.loc[df["ward"] == 9374, "ward_name"] = df.loc[df["ward"] == 9374, "ward_name"].fillna("Phường 2")
df.loc[df["ward"] == 9329, "ward_name"] = df.loc[df["ward"] == 9329, "ward_name"].fillna("Phường 3")
df.loc[df["ward"] == 9311, "ward_name"] = df.loc[df["ward"] == 9311, "ward_name"].fillna("Phường 9")
df.loc[df["ward"] == 9293, "ward_name"] = df.loc[df["ward"] == 9293, "ward_name"].fillna("Phường 3")
df.loc[df["ward"] == 9277, "ward_name"] = df.loc[df["ward"] == 9277, "ward_name"].fillna("Phường 11")
df.loc[df["ward"] == 9288, "ward_name"] = df.loc[df["ward"] == 9288, "ward_name"].fillna("Phường 6")
df.loc[df["ward"] == 9291, "ward_name"] = df.loc[df["ward"] == 9291, "ward_name"].fillna("Phường 2")

- Các căn hộ, chung cư sau sẽ được điền giá trị thiếu khi ta xem xét kĩ bằng cách tìm trên bản đồ và trang web.

In [14]:
df.loc[df["list_id"] == 85384567, "street_name"] = df.loc[df["list_id"] == 85384567, "street_name"].fillna("Đường Phạm Hùng")
df.loc[df["list_id"] == 85384567, "longitude"] = df.loc[df["list_id"] == 85384567, "longitude"].fillna(106.6834)
df.loc[df["list_id"] == 85384567, "latitude"] = df.loc[df["list_id"] == 85384567, "latitude"].fillna(10.7231)

df.loc[df["list_id"] == 35710979, "ward_name"] = df.loc[df["list_id"] == 35710979, "ward_name"].fillna("Phường 8")
df.loc[df["list_id"] == 35710979, "street_name"] = df.loc[df["list_id"] == 35710979, "street_name"].fillna("Đường Hoàng Diệu")
df.loc[df["list_id"] == 35710979, "longitude"] = df.loc[df["list_id"] == 35710979, "longitude"].fillna(106.7021)
df.loc[df["list_id"] == 35710979, "latitude"] = df.loc[df["list_id"] == 35710979, "latitude"].fillna(10.7627)

df.loc[df["list_id"] == 79821951, "street_name"] = df.loc[df["list_id"] == 79821951, "street_name"].fillna("Đường Mã Lò")
df.loc[df["list_id"] == 79821951, "longitude"] = df.loc[df["list_id"] == 79821951, "longitude"].fillna(106.6038)
df.loc[df["list_id"] == 79821951, "latitude"] = df.loc[df["list_id"] == 79821951, "latitude"].fillna(10.7669)

df.loc[df["list_id"] == 92978846, "street_name"] = df.loc[df["list_id"] == 92978846, "street_name"].fillna("Đường Tạ Quang Bửu")
df.loc[df["list_id"] == 92978846, "longitude"] = df.loc[df["list_id"] == 92978846, "longitude"].fillna(106.6688)
df.loc[df["list_id"] == 92978846, "latitude"] = df.loc[df["list_id"] == 92978846, "latitude"].fillna(10.7367)

df.loc[df["list_id"] == 35710979, "ward"] = df.loc[df["list_id"] == 35710979, "ward"].fillna(9400)
df.loc[df["list_id"] == 35710979, "toilets"] = df.loc[df["list_id"] == 35710979, "toilets"].fillna(2)

df.loc[df["list_id"] == 90729861, "street_name"] = df.loc[df["list_id"] == 90729861, "street_name"].fillna("Đường Mai Chí Thọ")
df.loc[df["list_id"] == 89829137, "street_name"] = df.loc[df["list_id"] == 89829137, "street_name"].fillna("Đường Phan Văn Trị")
df.loc[df["list_id"] == 88387767, "street_name"] = df.loc[df["list_id"] == 88387767, "street_name"].fillna("Đường Thân Nhân Trung")
df.loc[df["list_id"] == 91280266, "street_name"] = df.loc[df["list_id"] == 91280266, "street_name"].fillna("Đường Phan Văn Trị")
df.loc[df["list_id"] == 88289640, "street_name"] = df.loc[df["list_id"] == 88289640, "street_name"].fillna("Đường Hòa Bình")
df.loc[df["list_id"] == 74694531, "street_name"] = df.loc[df["list_id"] == 74694531, "street_name"].fillna("Đường Tôn Đức Thắng")
df.loc[df["list_id"] == 88197693, "street_name"] = df.loc[df["list_id"] == 88197693, "street_name"].fillna("Đường Bùi Văn Ba")
df.loc[df["list_id"] == 71029595, "street_name"] = df.loc[df["list_id"] == 71029595, "street_name"].fillna("Đường Tôn Đức Thắng")
df.loc[df["list_id"] == 93054197, "street_name"] = df.loc[df["list_id"] == 93054197, "street_name"].fillna("Đường Tạ Quang Bửu")

df.loc[df["list_id"] == 102494280, "price"] = df.loc[df["list_id"] == 102494280, "price"].fillna(16000000)
df.loc[df["list_id"] == 102494280, "price_million_per_m2"] = df.loc[df["list_id"] == 102494280, "price_million_per_m2"].fillna(0.21)

- Kiểm tra lại dữ liệu bị khuyết.

In [15]:
null_df = pd.DataFrame(df.isnull().sum(), columns=["Số mẫu bị thiếu"])
null_df["Tỉ lệ bị thiếu"] = np.round(df.isnull().sum() / df.shape[0] * 100, 2)
null_df.index.name = "Thuộc tính"
null_df = null_df.sort_values(by="Tỉ lệ bị thiếu", ascending=False)
null_df.loc[null_df["Tỉ lệ bị thiếu"] > 0]

Empty DataFrame
Columns: [Số mẫu bị thiếu, Tỉ lệ bị thiếu]
Index: []

## **2.10. Lưu dữ liệu**

In [16]:
df.to_csv("preprocessed_apartments.csv", index=False)